### PR TITLE
Add LCS functionality

### LINKED ISSUE #155: Implement remaining string commands
Extracted from [this comment](https://github.com/zio/zio-redis/issues/68#issuecomment-691777714) authored by @regis-leray.

### Commands

- [ ] STRALGO LCS algo-specific-argument [algo-specific-argument ...] Run algorithms (currently LCS) against strings

### Tips

- Define the API in `zio.redis.api`.
- Command-specific options should be defined in `zio.redis.options`.
- Additional inputs and outputs can be defined in `zio.redis.Input` and `zio.redis.Output`, respectively.
- Redis API evolves, therefore the commands listed above might be slightly changed. Double check [the official docs](https://redis.io/commands) before starting with implementation. 
- Use telnet to understand command protocol details.
- Write inputs and outputs tests in `zio.redis.InputSpec` and `zio.redis.OutputSpec`, respectively.
- Write integration tests by expanding the `zio.redis.ApiSpec`.
- If you feel that the pull request size is growing out of control, feel free to split it but make sure to link this issue in each of the related PRs.

### PR DESCRIPTION
**Pull request by barthorre**
Closes #155

### CODE REVIEW

File: redis/src/main/scala/zio/redis/options/Strings.scala [hunk 1]
```diff
@@ -2,6 +2,25 @@ package zio.redis.options
...
+
+  object StralgoCommand {
+    sealed case class StralgoLCS(`type`: StralgoType) extends StralgoCommand
+  }
+
+  sealed trait StralgoType { self =>
+    private[redis] final def stringify: String =
+      self match {
+        case StralgoType.Strings => "STRINGS"
+        case StralgoType.Keys    => "KEYS"
```
Comment (jxnu-liguobin) [inline redis/src/main/scala/zio/redis/options/Strings.scala:15]: I'm just looking around. Some options don't seem to be supported. List them? `[LEN] [IDX] [MINMATCHLEN <len>] [WITHMATCHLEN]
`

Comment (barthorre) [reply redis/src/main/scala/zio/redis/options/Strings.scala:15]: true, some of these would also alter the output, that's why I'd like some feedback on the input command structure before delving deeper into the other options.

File: redis/src/main/scala/zio/redis/options/Strings.scala [hunk 2]
```diff
@@ -2,6 +2,25 @@ package zio.redis.options
 
 trait Strings {
 
+  sealed trait StralgoCommand
+
+  object StralgoCommand {
+    sealed case class StralgoLCS(`type`: StralgoType) extends StralgoCommand
+  }
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/options/Strings.scala:9]: I'd avoid the command suffix as it's heavily used for redis commands. Additionally, perhaps it makes sense to unify the two, e.g.:

```scala
sealed trait Algorithm

object Algorithm {
  case object StringLcs extends Algorithm
  case object KeysLcs extends Algorithm
}
```

Of course, you can (and should) further expand with parameters as discussed by you and @jxnu-liguobin .

**Review by mijicd — COMMENTED** [review]
It goes in the right direction.

File: redis/src/main/scala/zio/redis/options/Strings.scala (hunk 2 above)
Comment (jxnu-liguobin) [reply redis/src/main/scala/zio/redis/options/Strings.scala:9]: it looks really good to use the unsuffixed one. It is more easy to illustrate the relationship of subcommands

File: redis/src/main/scala/zio/redis/options/Strings.scala [hunk 3]
```diff
@@ -2,6 +2,38 @@ package zio.redis.options
...
+  object StralgoLCS {
+    case object Strings extends StrAlgoLCS
+    case object Keys    extends StrAlgoLCS
+  }
+
+  sealed trait StrAlgoLcsQueryType
+
+  object StrAlgoLcsQueryType {
+    case object Len                                                           extends StrAlgoLcsQueryType
+    case class Idx(minMatchLength: Int = 1, withMatchLength: Boolean = false) extends StrAlgoLcsQueryType
```
Comment (anovakovic01) [inline redis/src/main/scala/zio/redis/options/Strings.scala:23]: It's not so clear from the documentation, but it looks like `MINMATCHLEN` can be specified without `IDX`.

Comment (barthorre) [reply redis/src/main/scala/zio/redis/options/Strings.scala:23]: it indeed seems like you can, but it doesn;t seem to do anything

image

Comment (anovakovic01) [reply redis/src/main/scala/zio/redis/options/Strings.scala:23]: I agree, so let's keep it where it already is.

File: redis/src/main/scala/zio/redis/options/Strings.scala (hunk 2 above)
Comment (barthorre) [reply redis/src/main/scala/zio/redis/options/Strings.scala]: I opted for the unsuffixed ones as that, to me, reflects the documentation/stucture of redis commands better.

**Review by anovakovic01 — APPROVED** [review]

**Review by mijicd — APPROVED** [review]

**Pull request merged**